source/io/module_eggShell.js

```diff
@@ -239,6 +246,257 @@
+            default:
+                throw new Error('Unexpected size for a Complex value');
+            }
+
+            validateVisitMethod(visitFn, 'visitComplex' + (sizeOfComplex === 8 ? 'Single' : 'Double'));
+            return visitFn.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitAnalogWaveform = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitAnalogWaveform, 'visitAnalogWaveform');
+            return typeVisitor.visitAnalogWaveform.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitTimestamp = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitTimestamp, 'visitTimestamp');
+            return typeVisitor.visitTimestamp.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitPath = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitPath, 'visitPath');
+            return typeVisitor.visitPath.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitArray = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitArray, 'visitArray');
+            return typeVisitor.visitArray.apply(typeVisitor, args);
+        };
+
+        var dispatchVisitCluster = function (typeVisitor, args) {
+            validateVisitMethod(typeVisitor.visitCluster, 'visitCluster');
+            return typeVisitor.visitCluster.apply(typeVisitor, args);
+        };
+
+        var typeDispatchFunctions = [
+            dispatchVisitBoolean,
+            dispatchVisitEnum,
+            dispatchVisitInteger,
+            dispatchVisitFloat,
+            dispatchVisitString,
+            dispatchVisitComplex,
+            dispatchVisitAnalogWaveform,
+            dispatchVisitTimestamp,
+            dispatchVisitPath,
+            dispatchVisitArray,
+            dispatchVisitCluster
+        ];
+
+        var registeredTypes = [
+            Module.eggShell.typeIsBoolean,
+            Module.eggShell.typeIsEnum,
+            Module.eggShell.typeIsInteger,
+            Module.eggShell.typeIsFloat,
+            Module.eggShell.typeIsString,
+            Module.eggShell.typeIsComplex,
+            Module.eggShell.typeIsAnalogWaveform,
+            Module.eggShell.typeIsTimestamp,
+            Module.eggShell.typeIsPath,
+            Module.eggShell.typeIsArray,
+            Module.eggShell.typeIsCluster
+        ];
+
+        Module.eggShell.reflectOnValueRef = publicAPI.eggShell.reflectOnValueRef = function (typeVisitor, valueRef) {
+            if (typeof valueRef !== 'object' || Array.isArray(valueRef)) {
+                throw new Error('valueRef must be an object. Found: ' + valueRef);
+            }
+
+            if (typeof typeVisitor !== 'object' || Array.isArray(typeVisitor)) {
+                throw new Error('typeVisitor must be an object. Found: ' + typeVisitor);
+            }
+
+            var typeRef = valueRef.typeRef,
+                args = Array.prototype.slice.call(arguments, 1),
```

Comment:
I think it's being a little too clever by letting users pass an arbitrary number of arguments. JS in general is faster when it can know ahead of time the structure of types being used.

So just having a `data` third parameter and then passing a
```
dispatcherParams = {
    type: typeRef,
    data: data
}
```
object to the dispatcher and in the dispatcher doing plain function invocation
```
typeVisitor.visitCluster(dispatcherParams.typeRef, dispatcherParams.data)
```

Reply:
Agreed, I'll change it